empress/support_files/js/biom-table.js

```diff
@@ -472,13 +472,102 @@ define(["underscore", "util"], function (_, util) {
+     * question. Keys in these objects are unique sample metadata values, and
+     * values in these objects are the proportion of samples containing the
+     * feature that have this unique value. Only frequency information for
+     * unique values where at least 1 sample with this value contains the
+     * feature is included in a given "frequencies" Object.
+     *
+     * This function is designed to be reasonably fast, which is a big part of
+     * why this works on the order of "each feature ID in the table" rather
+     * than on a feature-per-feature basis. (The reason for this design is that
+     * this is used for generating sample metadata barplots, and that was
+     * previously very slow on large trees: see issue #298 on GitHub. Thanks
+     * to Yoshiki for discussing this with me.)
+     *
+     * @param {String} col Sample metadata column
+     *
+     * @return {Object} fID2freqs
+     *
+     * @throws {Error} If the sample metadata column is unrecognized.
+     */
+    BIOMTable.prototype.getFrequencyMap = function (col) {
+        var scope = this;
+        var colIdx = this._getSampleMetadataColIndex(col);
+        var fIdx2counts = [];
+        var fIdx2sampleCt = [];
+        var containingSampleCount, cVal, cValIdx;
+
+        // Find unique (sorted) values in this sample metadata column; map
+        // sample metadata values to a consistent index. (Using an index to
+        // store this data means we can store the sample metadata values for
+        // each feature in an Array rather than in an Object for now.)
+        var uniqueSMVals = this.getUniqueSampleValues(col);
+        var numUniqueSMVals = uniqueSMVals.length;
+        var smVal2Idx = {};
+        _.each(uniqueSMVals, function (smVal, c) {
+            smVal2Idx[smVal] = c;
+        });
+
+        // Assign each feature an empty counts array with all 0s. Also set
+        // things up so we can keep track of the total number of samples
+        // containing each feature easily.
+        var i, emptyCounts;
+        _.each(this._fIDs, function (fID, fIdx) {
+            emptyCounts = [];
+            for (i = 0; i < numUniqueSMVals; i++) {
+                emptyCounts.push(0);
+            }
+            fIdx2counts.push(emptyCounts);
+            fIdx2sampleCt.push(0);
+        });
+
+        // Iterate through each sample of the BIOM table, storing unique s.m.
+        // value counts and total sample counts for each feature
+        _.each(this._tbl, function (presentFeatureIndices, sIdx) {
+            // Figure out what metadata value this sample has at the column.
+            cVal = scope._sm[sIdx][colIdx];
+            cValIdx = smVal2Idx[cVal];
+            // Increment s.m. value counts for each feature present in this
+            // sample
+            _.each(presentFeatureIndices, function (fIdx) {
+                fIdx2counts[fIdx][cValIdx]++;
+                fIdx2sampleCt[fIdx]++;
+            });
```

Comment:
```suggestion
        _.each(this._tbl, function (samples, sIdx) {
            // Figure out what metadata value this sample has at the column.
            cVal = scope._sm[sIdx][colIdx];
            cValIdx = smVal2Idx[cVal];
            // Increment s.m. value counts for each feature present in this
            // sample
            _.each(samples, function (fIdx) {
                fIdx2counts[fIdx][cValIdx]++;
                fIdx2sampleCt[fIdx]++;
            });
```

Since _tbl is a 2D array of samples by features, this may make more sense to developers who are not to familiar with biom-table. Minor suggestion feel free to ignore. (:

Reply:
If it's ok I'd like to keep this as is -- there are a couple of other places in the BIOM table JS where `presentFeatureIndices` is used when iterating over the table in this way. I added some context to the comment before this loop, so hopefully that makes things clearer as a compromise :) You're totally correct, though -- we should make this code as non-intimidating as possible, since these functions are all getting used pretty frequently ...